src/main/kotlin/org/rust/ide/inspections/RsSortImplTraitMembersInspection.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.project.Project
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsItemElement
+import org.rust.lang.core.psi.ext.RsTraitOrImpl
+import org.rust.lang.core.psi.ext.elementType
+import org.rust.lang.core.psi.ext.resolveToTrait
+
+class RsSortImplTraitMembersInspection : RsLocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = object : RsVisitor() {
+        override fun visitImplItem(implItem: RsImplItem) {
+            val traitItems = targetItems(implItem.traitRef?.resolveToTrait).takeIf { it.isNotEmpty() } ?: return
+            val implItems = targetItems(implItem).takeIf { it.isNotEmpty() } ?: return
+            if (traitItems.size != implItems.size) return
+            if (traitItems.zip(implItems).all { it.first.isSameItem(it.second) }) return
+            holder.registerProblem(
+                implItem,
+                "Different impl member order from the trait",
+                SortImplTraitMembersFix(traitItems, implItems)
+            )
+        }
+    }
+
+    private fun targetItems(item: RsTraitOrImpl?): List<RsItemElement> {
+        val children = item?.members?.children
+        return children?.mapNotNull {
+            if (it is RsFunction || it is RsConstant || it is RsTypeAlias) it as? RsItemElement else null
+        }?.takeIf { it.size == children.size } ?: emptyList()
+    }
+
+    private class SortImplTraitMembersFix(
+        private val traitItems: List<RsItemElement>,
+        private val implItems: List<RsItemElement>
```

Comment:
Let's avoid holding psi elements in quick fix object - it can lead to memory leak.
Usually, if you want to use some psi element in quick fix, you have several options:
* use `com.intellij.psi.SmartPsiElementPointer`
* get psi element from `ProblemDescriptor` in `applyFix` method. You get same psi element that you use in `registerProblem`
* use `com.intellij.codeInspection.LocalQuickFixOnPsiElement` as parent class of your quick fix

2nd and 3rd options use `com.intellij.psi.SmartPsiElementPointer` under the hood so basically there is only one way to do right.

I think here we can use `LocalQuickFixOnPsiElement` to pass impl and trait objects, create separate method to get items from them and check all conditions (like `findApplicableContext` in `RsElementBaseIntentionAction`) and use this method in `visitImplItem` and in `applyFix`.